### ISSUE TITLE
Remove defaults only after validation has completed

### ISSUE THREAD

**Issue by eslavich** [spacetelescope/asdf#795]
Astropy [encountered](https://github.com/astropy/astropy/issues/10327) this bug introduced by #793.  Validation was combined with removing defaults in such a way that the order of those operations was not preserved.  Now removing defaults occurs first, which leads to validation errors when the removed value is required by the schema.

We may be able to rework the custom properties validator, but if not, it's probably okay to separate removing defaults and validation again, since that only occurs on write and I think the goal of the effort was to improve read performance.  In any case we should add a test to cover this behavior.

**Comment by eslavich** [comment]
I looked into the astropy failures more carefully, and discovered problems with filling defaults as well.  The issue comes when filling defaults for a schema with an `anyOf` or `oneOf` combiner -- defaults get filled for subschemas that may not actually match the object.  That's difficult to avoid, and we've started discussing dropping the defaults feature entirely, so I'm just going to revert the change that caused this.
